src/container/container_iv.c

```diff
@@ -205,35 +156,58 @@ cont_iv_ent_destroy(d_sg_list_t *sgl)
 }
 
 static int
-cont_iv_ent_copy(struct ds_iv_entry *entry, d_sg_list_t *dst_sgl,
-		 struct cont_iv_entry *src)
+cont_iv_ent_copy(struct ds_iv_entry *entry, struct cont_iv_key *key,
+		 d_sg_list_t *dst_sgl, struct cont_iv_entry *src)
 {
-	struct cont_iv_entry *dst = dst_sgl->sg_iovs[0].iov_buf;
+	struct cont_iv_entry	*dst = dst_sgl->sg_iovs[0].iov_buf;
+	daos_size_t		size;
+	uint64_t		snap_cnt;
+	int			rc = 0;
 
 	uuid_copy(dst->cont_uuid, src->cont_uuid);
 	switch (entry->iv_class->iv_class_id) {
 	case IV_CONT_SNAP:
-		D_ASSERT(src->iv_snap.snap_cnt <= MAX_SNAP_CNT);
+		if (src->iv_snap.snap_cnt == (uint64_t)(-1)) {
+			snap_cnt = 1;
+			rc = -DER_IVCB_FORWARD;
+		} else {
+			snap_cnt = src->iv_snap.snap_cnt;
+		}
+
+		D_DEBUG(DB_MD, "snap_cnt "DF_U64":"DF_U64"\n",
+			snap_cnt, src->iv_snap.snap_cnt);
+		size = cont_iv_snap_ent_size(snap_cnt);
+		if (size > dst_sgl->sg_iovs[0].iov_buf_len) {
+			/* Return -1 so the client fetch will reallocate
+			 * the buffer.*/
+			dst->iv_snap.snap_cnt = (uint64_t)-1;
+			dst->iv_snap.snaps[0] = src->iv_snap.snap_cnt;
```

Comment:
I don't see why we have to set -1 to 'snap_cnt', why don't directly set the requested number in 'snap_cnt'?

Reply:
It needs a way to tell if the reply is for requesting more snapshot buffer, or it is just correct reply.  And error code itself can not pack extra information, that is why it use -1 here.

Reply:
Then that will add rc for every case. even with the value inside iv_entry, which actually does not make things clearer.